# HG changeset patch
# User coffeys
# Date 1535365754 -3600
#      Mon Aug 27 11:29:14 2018 +0100
# Node ID 8796258816cd385a3763e1b64b494e6328de0f6e
# Parent  b63702f5cc12b308770ec4bb622d614a7fc3b30b
8208585: Make crypto code more robust
Reviewed-by: ascarpino, mschoene

diff -r b63702f5cc12 -r 8796258816cd src/share/classes/com/sun/crypto/provider/RSACipher.java
--- openjdk/jdk/src/share/classes/com/sun/crypto/provider/RSACipher.java	Thu Aug 23 13:23:39 2018 +0530
+++ openjdk/jdk/src/share/classes/com/sun/crypto/provider/RSACipher.java	Mon Aug 27 11:29:14 2018 +0100
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -329,7 +329,7 @@
         if ((inLen == 0) || (in == null)) {
             return;
         }
-        if (bufOfs + inLen > buffer.length) {
+        if (inLen > (buffer.length - bufOfs)) {
             bufOfs = buffer.length + 1;
             return;
         }
diff -r b63702f5cc12 -r 8796258816cd src/share/classes/sun/security/pkcs11/P11Signature.java
--- openjdk/jdk/src/share/classes/sun/security/pkcs11/P11Signature.java	Thu Aug 23 13:23:39 2018 +0530
+++ openjdk/jdk/src/share/classes/sun/security/pkcs11/P11Signature.java	Mon Aug 27 11:29:14 2018 +0100
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -472,6 +472,10 @@
         if (len == 0) {
             return;
         }
+        // check for overflow
+        if (len + bytesProcessed < 0) {
+            throw new ProviderException("Processed bytes limits exceeded.");
+        }
         switch (type) {
         case T_UPDATE:
             try {
diff -r b63702f5cc12 -r 8796258816cd src/share/classes/sun/security/provider/DSA.java
--- openjdk/jdk/src/share/classes/sun/security/provider/DSA.java	Thu Aug 23 13:23:39 2018 +0530
+++ openjdk/jdk/src/share/classes/sun/security/provider/DSA.java	Mon Aug 27 11:29:14 2018 +0100
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1996, 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -491,7 +491,7 @@
                 }
             }
             protected void engineUpdate(byte[] input, int offset, int len) {
-                if (ofs + len > digestBuffer.length) {
+                if (len > (digestBuffer.length - ofs)) {
                     ofs = Integer.MAX_VALUE;
                 } else {
                     System.arraycopy(input, offset, digestBuffer, ofs, len);
@@ -500,7 +500,7 @@
             }
             protected final void engineUpdate(ByteBuffer input) {
                 int inputLen = input.remaining();
-                if (ofs + inputLen > digestBuffer.length) {
+                if (inputLen > (digestBuffer.length - ofs)) {
                     ofs = Integer.MAX_VALUE;
                 } else {
                     input.get(digestBuffer, ofs, inputLen);
diff -r b63702f5cc12 -r 8796258816cd src/windows/classes/sun/security/mscapi/RSASignature.java
--- openjdk/jdk/src/windows/classes/sun/security/mscapi/RSASignature.java	Thu Aug 23 13:23:39 2018 +0530
+++ openjdk/jdk/src/windows/classes/sun/security/mscapi/RSASignature.java	Mon Aug 27 11:29:14 2018 +0100
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2005, 2012, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2018, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -132,7 +132,7 @@
         @Override
         protected void engineUpdate(byte[] b, int off, int len)
                 throws SignatureException {
-            if (offset + len > precomputedDigest.length) {
+            if (len > (precomputedDigest.length - offset)) {
                 offset = RAW_RSA_MAX + 1;
                 return;
             }
@@ -147,7 +147,7 @@
             if (len <= 0) {
                 return;
             }
-            if (offset + len > precomputedDigest.length) {
+            if (len > (precomputedDigest.length - offset)) {
                 offset = RAW_RSA_MAX + 1;
                 return;
             }
